mcproto/types/nbt.py

```diff
@@ -0,0 +1,1315 @@
+                    raise ValueError("Integer list contains values out of range.")
+                return LongArrayNBT(data, name=name)
+            return ListNBT([NBTag.from_object(item, use_int_array=use_int_array) for item in data], name=name)
+        if isinstance(data, dict):
+            if len(data) == 0:
+                return CompoundNBT([], name=name)
+            if len(data) == 1 and name == "":
+                key, value = next(iter(data.items()))
+                return NBTag.from_object(value, name=key, use_int_array=use_int_array)
+            payload = []
+            for key, value in data.items():
+                tag = NBTag.from_object(value, name=key, use_int_array=use_int_array)
+                payload.append(tag)
+            return CompoundNBT(payload, name)
+        if data is None:
+            warnings.warn("Converting None to an END tag.", stacklevel=2)
+            return EndNBT()  # Should not be used
+
+        try:
+            # Check if the object can be converted to bytes
+            return ByteArrayNBT(bytes(data), name=name)  # type: ignore
+        except (TypeError, ValueError):
+            pass
+        raise TypeError(f"Cannot convert object of type {type(data)} to an NBT tag.")
+
+    def to_object(self) -> Mapping[str, PayloadType] | PayloadType:
+        """Convert the NBT payload to a dictionary."""
+        return CompoundNBT(self.payload).to_object()  # allow NBTag.to_object to act as a dict
+
+    def __getitem__(self, key: str | int) -> PayloadType:
+        """Get a tag from the list or compound tag."""
+        if self.TYPE not in (NBTagType.LIST, NBTagType.COMPOUND, NBTagType.INT_ARRAY, NBTagType.LONG_ARRAY):
+            raise TypeError(f"Cannot get a tag by index from a non-LIST or non-COMPOUND tag ({self.TYPE}).")
+
+        if not isinstance(self.payload, list):
+            raise AttributeError(
+                f"The payload of the tag is not a list ({self.TYPE}).\n"
+                "Check that the initialization of the tag is correct."
+            )
+        if not isinstance(key, (str, int)):  # type: ignore
+            raise TypeError("Key must be a string or an integer.")
+
+        if isinstance(key, str):
+            if self.TYPE != NBTagType.COMPOUND:
+                raise TypeError(f"Cannot get a tag by name from a non-COMPOUND tag ({self.TYPE}).")
+            if not all(isinstance(tag, NBTag) for tag in self.payload):
+                raise AttributeError("The payload of the tag is not a list of NBTag objects.")
+            for tag in self.payload:
+                tag = cast(NBTag, tag)
+                if tag.name == key:
+                    return tag
+            raise KeyError(f"No tag with the name {key!r} found.")
+
+        # Key is an integer
+        if key < -len(self.payload) or key >= len(self.payload):
+            raise IndexError(f"Index {key} out of range.")
+        return self.payload[key]
```

Comment:
wouldn't `self.payload[key]` cause this exception anyway? why handle it ourselves here?